### PR TITLE
[AMDGPU] Create local KnownBits in case DenseMap gets invalidated

### LINKED ISSUE #110930: [AMDGPU] Invalid integer comparison assertion hit
The following code hits an assertion, this case only appeared when I compiled with remarks enabled for some reason. https://godbolt.org/z/6v5s7aPs7. Without assertions it will successfully build.

### PR DESCRIPTION
**Pull request by JanekvO**
KnownBits retrieved from DenseMap may invalidate if insertion requires a (re)growth.

Fixes https://github.com/llvm/llvm-project/issues/110930

### CODE REVIEW

File: llvm/test/CodeGen/AMDGPU/mcexpr-knownbits-assign-crash.ll [hunk 1]
```diff
@@ -0,0 +1,120 @@
+; REQUIRES: asserts
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/mcexpr-knownbits-assign-crash.ll:1]: It does not require asserts

File: llvm/test/CodeGen/AMDGPU/mcexpr-knownbits-assign-crash.ll [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+; RUN: llc -mtriple=amdgcn-amd-amdhsa -mcpu=gfx1030 -o - %s 2>&1 | FileCheck %s
+
+; CHECK-NOT: Assertion `BitWidth == RHS.BitWidth && "Bit widths must be same for comparison"' failed
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/mcexpr-knownbits-assign-crash.ll:4]: A check-not of an assert message is next to worthless, positively check the output

File: llvm/test/CodeGen/AMDGPU/mcexpr-knownbits-assign-crash.ll [hunk 3]
```diff
@@ -0,0 +1,120 @@
+; REQUIRES: asserts
+; RUN: llc -mtriple=amdgcn-amd-amdhsa -mcpu=gfx1030 -o - %s 2>&1 | FileCheck %s
+
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/mcexpr-knownbits-assign-crash.ll:3]: Add a comment explaining what this tests. Include the issue number in the filename?

File: llvm/test/CodeGen/AMDGPU/mcexpr-knownbits-assign-crash-gh-issue-110930.ll [hunk 4]
```diff
@@ -1,120 +1,144 @@
...
-define i32 @R_FlatNumForName() {
-  call void (...) @I_Error()
-  unreachable
+; CHECK-LABEL: RunTic:
+define void @RunTic() {
+  %call5.i1 = call i32 @G_CheckDemoStatus()
+  tail call void @D_AdvanceDemo()
+  call void @G_Ticker()
+  ret void
 }
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/mcexpr-knownbits-assign-crash-gh-issue-110930.ll:144]: Check resource limits?

File: llvm/test/CodeGen/AMDGPU/mcexpr-knownbits-assign-crash-gh-issue-110930.ll [hunk 5]
```diff
@@ -1,120 +1,144 @@
...
+define void @P_RemoveMobj() {
   ret void
 }
 
-define void @G_DoReborn() {
-  call void @P_RemoveMobj()
-  call void @P_SpawnMobj()
-  call void @P_SpawnPlayer()
-  call void (...) @I_Error()
+; CHECK-LABEL: P_SpawnMobj:
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/mcexpr-knownbits-assign-crash-gh-issue-110930.ll:20]: Just checking all the labels isn't really useful

**Comment by JanekvO** [comment]
Rebase

**Review by arsenm — APPROVED** [review]

**Pull request merged**